File: code/08/ruby/project/Temperature.rb
#Temperature.rb is a class for Temperature
#
#Created by: Mark Davis, CS 214, CALVIN COLLEGE
#Date: April 12, 2016
#project08
##############################################
module Temperatures

class Temperature

  #####################################################
  # initialize() constructs Temperature instance variables
  # POST: myDegree and myScale have values
  #####################################################
  def initialize(degree, scale)
	if isValid(degree, scale)   #check if valid
		@myDegree = degree #then set instance variables
		@myScale = scale
	end
  end

  #####################################################
  # isValid() checks valid scales and temperatures
  # Return: true if valid, false if not valid
  #####################################################
  def isValid(degree, scale)
	if scale == "k" or scale == "K" and degree < 0.0 #validity for Kelvin
		return false
	elsif scale == "c" or scale == "C" and degree < -237.0 #validity for Celsius
		return false
	elsif scale == "f" or scale == "F" and degree < -459.0 #validity for F
		return false
	else
		return true
	end
  end


  #####################################################
  # attr_reader is accessor for instance variables
  # Return: myDegree or myScale
  #####################################################
  attr_reader :myDegree, :myScale #accessors for myDegree and myScale

  #####################################################
  # setCelsius() sets values to Celsius from other scales
  # POST: myDegree has new value, myScale has new value
  #####################################################
  def setCelsius
	if @myScale == 'K' or @myScale == 'k'
		@myDegree = @myDegree - 273.15 #conversion equation
		@myScale = 'C'
	elsif @myScale == 'F' or @myScale == 'f'
		@myDegree = (@myDegree - 32.0) * (5.0/9.0) #conversion equation
		@myScale = 'C'
	end
  end

  #####################################################
  # setKelvin() sets values to Kelvin from other scales
  # POST: myDegree has new value, myScale has new value
  #####################################################
  def setKelvin
	if @myScale == 'C' or @myScale == 'c'
		@myDegree = @myDegree + 273.15 #conversion equation
		@myScale = 'K'
	elsif @myScale == 'F' or @myScale == 'f'
		@myDegree = (@myDegree + 459.67) * (5.0/9.0) #conversion equation
		@myScale = 'K'
	end
  end

  #####################################################
  # setFahrenheit() sets values to F from other scales
  # POST: myDegree has new value, myScale has new value
  #####################################################
  def setFahrenheit
	if @myScale == 'C' or @myScale == 'c'
		@myDegree = @myDegree * (9.0/5.0) + 32.0 #conversion equation
		@myScale = 'F'
	elsif @myScale == 'K' or @myScale == 'k'
		@myDegree = @myDegree * (9.0/5.0) - 459.67 #conversion equation
		@myScale = 'F'
	end
  end

  #####################################################
  # raise() raises value of myDegree
  # POST: myDegree has new value
  #####################################################
  def raise(addition)
	@myDegree = @myDegree + addition #add on new temperature
  end

  #####################################################
  # lower() lowers value of myDegree
  # POST: myDegree has new value
  #####################################################
  def lower(subtraction)
	@myDegree = @myDegree - subtraction #add on new temperature
  end

  #####################################################
  # equals() compares two temperatures for equality
  # returns true if equal, false if not equal
  #####################################################
  def equals(deg2)
	result = false
	if @myScale == deg2.myScale #if scale is equal
		if @myDegree ==	deg2.myDegree #if degree is equal
			result = true
		end
	else
		if @myScale == 'K' or @myScale == 'k' #if Kelvin is scale
			deg2.setKelvin
		elsif @myScale == 'C' or @myScale == 'c' #if scale is Celsius
			deg2.setCelsius
		elsif @myScale == 'F' or @myScale == 'f' #if scale is F
			deg2.setFahrenheit
		end
		if @myDegree == deg2.myDegree #if degrees are equal
			result true
		end
	end
	return result	
  end

  #####################################################
  # lessThan() compares two temperatures for less than
  # returns true if less, false if not less
  #####################################################
  def lessThan(deg2)
	result = false
	if @myScale == deg2.myScale #if degrees are equal
		if @myDegree <	deg2.myDegree #if degrees are equal
			result = true
		end
	else
		if @myScale == 'K' or @myScale == 'k' #if scale is Kelvin
			deg2.setKelvin
		elsif @myScale == 'C' or @myScale == 'c' #if scale is Celsius
			deg2.setCelsius
		elsif @myScale == 'F' or @myScale == 'f' #if scale is F
			deg2.setFahrenheit
		end
		if @myDegree < deg2.myDegree #if less than
			result = true
		end
	end
	return result
  end

  #####################################################
  # readTemp() reads in a Temperature from user
  # POST: Temperature has new scale and new degree
  #####################################################
  def readTemp
	#get input for degree
	puts "Enter degree: "
	degree = gets.chomp.to_f #convert to a float
	@myDegree = degree
	#get input for scale
	puts "Enter scale: "
	scale = gets.chomp #convert to a char
	@myScale = scale
  end

  #####################################################
  # printTemp() prints out a Temperature
  # POST: Temperature has new scale and new degree
  #####################################################
  def printTemp
	"#{myDegree} #{myScale}" #return on one line
  end

end
end

File: code/08/java/Temperature.java
/* temperature.java implements class temperature.
 *
 * Completed by: Mark Davis
 * Date: April 12, 2016
 */

import java.util.Scanner;
class Temperature
{

    private double myDegree;
    private char myScale; 

    /**************************************************
     * Temperature constructs a temperature from and int and string*
     * Receive: degrees and scale. *
     * Return: the double (myDegree, myScale).        *
     ***************************************************/  
    public Temperature(char scale, double degrees)
	{
	if ( isValid(scale, degrees)) {
		myDegree = degrees;
		myScale = scale;	
		}
	}

    /**************************************************
     * isValid returns if a temperature is valid or not
     * Receive: degrees and scale.                    *
     * Return: true or false                          *
     ***************************************************/
    private boolean isValid(char scale, double degrees) {
	boolean result = true;
	if(scale == 'K' || scale == 'k' && degrees < 0.0){
		result = false;
	}
	else if(scale == 'C' || scale == 'c' && degrees < -237.0) {
		result = false;
	}
	else if(scale == 'F' || scale == 'f' && degrees < -459.0) {
		result = false;
	}
	else {
		result = true;
 	}
	return result;
	}

    /***************************************************
     * getDegrees() extracts the degrees of the temperature*
     * Return: myDegree                                *
     ***************************************************/
    public double getDegree() 
	{
	return myDegree;
	}

    /***************************************************
     * getScale() extracts the scale of the temperature*
     * Return: myScale.                                *
     ***************************************************/
    public char getScale()
	{
	return myScale;
	}

    /***************************************************
     * getFahrenheit() returns temperature in fahrenheit*
     * Return: myDegree.                                *
     ***************************************************/
    public void getFahrenheit()
	{
	if(myScale == 'C' || myScale == 'c') {
		myDegree = myDegree * (9f/5) + 32.0;
		myScale = 'F'; }
	if(myScale == 'K' || myScale == 'k') {
		myDegree = myDegree * (9f/5) - 459.67;
		myScale = 'F'; }
	}

    /***************************************************
     * getCelsius() returns temperature in Celsius     *
     * Return: myDegree.                               *
     ***************************************************/
    public void getCelsius()
	{
	if(myScale == 'F' || myScale == 'f') {
		myDegree = (myDegree - 32.0) * (5f/9);
		myScale = 'C'; }
	if(myScale == 'K' || myScale == 'k') {
		myDegree = myDegree - 273.15;
		myScale = 'C'; }
	}

    /***************************************************
     * getKelvin() returns temperature in Kelvin       *
     * Return: myDegree.                               *
     ***************************************************/
    public void getKelvin()
	{
	if(myScale == 'F' || myScale == 'f') {
		myDegree = (myDegree + 459.67) * (5f/9);
		myScale = 'K'; }
	if(myScale == 'C' || myScale == 'c') {
		myDegree = myDegree + 273.15;
		myScale = 'K'; }
	}

    /***************************************************
     * raise() raises a temperature                    *
     * POST: new temperature is set                    *
     ***************************************************/
    void raise(double addition) {
	myDegree = myDegree + addition;
	}

    /***************************************************
     * lower() lowers a temperature                    *
     * POST: new temperature is set                    *
     ***************************************************/
    void lower(double myDegrees, double subtraction) {
		myDegree = myDegree - subtraction;
	}

	
    /***************************************************
     * equals() compares two temperatures              *
     * Return: true if equal, false if not equal       *
     ***************************************************/
    public boolean equals(Temperature deg2) {
	boolean result = false;
	if(myScale == deg2.getScale()) {
		if (myDegree == deg2.getDegree()) {
			result = true;
		}
	}
	else {
		if (myScale == 'k' || myScale == 'K') {
			deg2.myScale = 'K';}
		else if (myScale == 'C' || myScale == 'c') {
			deg2.myScale = 'C';}
		else if (myScale == 'F' || myScale == 'f') {
			deg2.myScale = 'F'; }
		if (myDegree == deg2.getDegree()) {
			result = true; }	
	}	
	return result;
	}

   /***************************************************
     * lessThan() determines if a temp is less than another temp*
     * Return: true if less, false if not less         *
     ***************************************************/
    public boolean lessThan(Temperature deg2) {
	boolean result = false;
	if(myScale == deg2.getScale()) {
		if (myDegree < deg2.getDegree()) {
			result = true;
		}
	}
	else {
		if (myScale == 'k' || myScale == 'K') {
			deg2.getKelvin(); }
		else if (myScale == 'C' || myScale == 'c') {
			deg2.getCelsius(); }
		else if (myScale == 'F' || myScale == 'f') {
			deg2.getFahrenheit(); }
		if (myDegree < deg2.getDegree()) {
			result = true; }	
	}	
	return result;
    }
	
    /***************************************************
     * readTemp() gets user input for a Temperature    *
     * POST: new temperature set from user             *
     ***************************************************/
    void readTemp() {
	Scanner keyboard = new Scanner(System.in);
	//get temperature
	System.out.println("Enter degrees: ");
	double degree = keyboard.nextDouble();
	myDegree = degree;
	
	//get scale
	System.out.println("Enter scale: ");
	char scale = keyboard.next().charAt(0);
	myScale = scale;
	}


    /***************************************************
     * printTemp() returns Temperature type            *
     * Return: myDegree and myScale                    *
     ***************************************************/
    public String printTemp() {
	return String.valueOf(myDegree + " " + String.valueOf(myScale));
	}
}
